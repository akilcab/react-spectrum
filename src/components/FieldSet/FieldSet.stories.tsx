import { Story, Meta } from "@storybook/react/types-6-0";
import { Form, Formik, FormikProps, FormikValues } from "formik";
import { FieldSet } from "./FieldSet";

export default {
  title: "FieldSet",
  component: FieldSet,
  description: `A FieldSet.`,
  argTypes: {},
} as Meta;

//👇 We create a “template” of how args map to rendering
const Template: Story = (args) => (
  <Formik initialValues={{ email: "" }} onSubmit={() => {}}>
    {({ handleChange, isSubmitting }: FormikProps<FormikValues>) => (
      <Form>
        <FieldSet
          name={"email"}
          placeholder={"Entrez votre email"}
          type={"email"}
          {...args}
        />
      </Form>
    )}
  </Formik>
);

//👇 Each story then reuses that template
export const Default = Template.bind({});
Default.args = {};
