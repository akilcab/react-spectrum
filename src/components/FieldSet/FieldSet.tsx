import React from "react";
import FormHelperText from "@material-ui/core/FormHelperText/FormHelperText";
import { ErrorMessage, Field, FieldProps } from "formik";
import { InputBaseProps } from "@material-ui/core";
import _ from "lodash";
import { BootstrapInput } from "./components/BootstrapInput";
import { GlobalStyles } from "../../utils/styles";

export interface FieldSetProps extends InputBaseProps {
  type?: string;
  name: string;
  placeholder: string;
  onChange?: (e: React.ChangeEvent<any>) => void;
  value?: string | number | boolean | null;
  component?: React.ComponentType;
  as?: string | React.ComponentType<FieldProps["field"]>;
  options?: any[];
  disabled?: boolean;
}

export const FieldSet: React.FC<FieldSetProps> = (props) => {
  const {
    name,
    placeholder,
    type,
    value,
    disabled,
    inputProps,
    ...other
  } = props;
  const classes = GlobalStyles();

  return (
    <Field name={name} type={type}>
      {({ field, form: { touched, errors } }: FieldProps) => {
        const touch = _.get(touched, name);
        const err = _.get(errors, name);

        return (
          <>
            <BootstrapInput
              {...other}
              placeholder={placeholder}
              {...field}
              value={value}
              error={touch && !!err}
              type={type}
              disabled={disabled}
            />
            <ErrorMessage
              name={name}
              component={FormHelperText}
              className={classes.formHelperText}
            />
          </>
        );
      }}
    </Field>
  );
};