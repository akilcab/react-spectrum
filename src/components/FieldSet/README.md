# Description of component

contains as few styles as possible. It aims to be a simple building block for creating an input. It contains a load of
style reset and some state logic.

#### Example :

![](assets/desc.png)

## How to import

```node
import { FieldSet } from "app/shared/components/FieldSet";
```

## How to use

```node
<Formik>
  {({ handleChange }) => (
    ....
      <FieldSet
        type="email"
        name="email"
        onChange={handleChange}
        placeholder="Entrez votre email"
      />
    ....
  )}
<Formik />
```

## Props

All others InputBase props

## CSS properties we can override

You can override all CSS properties of the InputBaseProps type of Material UI.
