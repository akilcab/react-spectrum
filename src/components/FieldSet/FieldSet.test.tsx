import { Form, Formik } from "formik";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import { FieldSet } from "./FieldSet";

const handleSubmit = jest.fn();

describe("<FieldSet />", () => {
  test("rend correctement", () => {
    const component = renderer
      .create(
        <Formik
          initialValues={{ email: "" }}
          onSubmit={handleSubmit}
          validationSchema={{}}
        >
          <Form>
            <FieldSet name="email" placeholder="Entrez votre email" />
          </Form>
        </Formik>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test("should have not content texts in the component", () => {
    const { container } = render(
      <Formik
        initialValues={{ email: "" }}
        onSubmit={handleSubmit}
        validationSchema={{}}
      >
        <Form>
          <FieldSet name="name" placeholder="Entrez votre email" />
        </Form>
      </Formik>
    );
    expect(container).not.toHaveTextContent("Entrez votre email");
  });

  test("try to display the profile without its require props", () => {
    expect(FieldSet).toThrowError();
  });
});
