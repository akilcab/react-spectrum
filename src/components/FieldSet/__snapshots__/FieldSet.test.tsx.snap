// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldSet /> rend correctement 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiInputBase-root WithStyles(ForwardRef(InputBase))-root-17"
    onClick={[Function]}
  >
    <input
      className="MuiInputBase-input WithStyles(ForwardRef(InputBase))-input-18 WithStyles(ForwardRef(InputBase))-input-19"
      name="email"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Entrez votre email"
      type="text"
    />
  </div>
</form>
`;
